File: Online Judge Problems/Competitive Programming 4/Mathematics/Binomial Coefficients/combinations.cpp
// combinations (https://onlinejudge.org/index.php?option=com_onlinejudge&Itemid=8&category=24&page=show_problem&problem=305)
 
#include <iostream>
#include <vector>
 
using namespace std;
 
// Time: O(N * K)
// Space: O(N * K)
 
using ll = long long;

ll binomialCoefficient(ll n, ll k, vector<vector<ll>>& dp) {
    if (k == 0 || k == n)
        return 1;
    if (dp[n][k] != -1)
        return dp[n][k];
    if (k > n-k)
        k = n-k;
    
    int included = binomialCoefficient(n-1, k-1, dp);
    int excluded = binomialCoefficient(n-1, k, dp);

    return dp[n][k] = included + excluded;
}

void solve(ll obj_cnt, ll choose_cnt) {
    vector<vector<ll>> dp(obj_cnt+1, vector<ll>(choose_cnt+1, -1));

    ll coefficient = binomialCoefficient(obj_cnt, choose_cnt, dp);
    cout << obj_cnt << " things taken " << choose_cnt << " at a time is ";
    cout << coefficient << " exactly." << endl;
}
 
int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
       
    ll obj_cnt, choose_cnt;
    cin >> obj_cnt >> choose_cnt;

    while (obj_cnt > 0 && choose_cnt > 0) {
        solve(obj_cnt, choose_cnt);
        
        cin >> obj_cnt >> choose_cnt;
    }
    
    return 0;
}

File: Online Judge Problems/Competitive Programming 4/Mathematics/Binomial Coefficients/combinations.py
# Combinations (https://onlinejudge.org/index.php?option=com_onlinejudge&Itemid=8&category=24&page=show_problem&problem=305)

def factorial(x, dp):
    if x == 0:
        return 1
    if dp.get(x, -1) != -1:
        return dp[x]
    dp[x] = x * factorial(x-1, dp)
    return dp[x]

def bin_coeff(n, k, dp):
    numer = factorial(n, dp)
    denom = factorial(n-k, dp) * factorial(k, dp)
    return numer // denom

def main():
    dp = {}
    n, k = map(int, input().split(' '))

    while n > 0 and k > 0:
        coeff = bin_coeff(n, k, dp)
        print(f"{n} things taken {k} at a time is {coeff} exactly.")

        n, k = map(int, input().split(' '))
    

if __name__ == "__main__":
    main()